### PR TITLE
core: load node part of kotlin infra directly from railjson

### PR DESCRIPTION
**Pull request by bougue-pe**
Also:
* add converter from EdgeEndpoint (railjson) in Endpoint (kt).
* add comparison tests on node part of the resulting infra
* cleanup in first 2 commits
* update test to use new builder's interface.
* continues on track-section's directions:
  * drop more parts of RawInfraBuilder interface.
  * maintain stitching to connect "new" loading of part of the infra and "old" loading of the rest of the infra.

Tested locally on France and Germany infra ✅
🔍 better reviewed by commit with message

Part of https://github.com/osrd-project/osrd-confidential/issues/307

### CODE REVIEW

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt [hunk 1]
```diff
@@ -84,31 +86,49 @@ class RawInfraFromRjsBuilderImpl : RawInfraBuilder {
...
-        movableElementBuilder.init()
-        val movableElement = movableElementBuilder.build()
-        return trackNodePool.add(movableElement)
+        val nodeIdx = trackNodePool.add(TrackNodeDescriptor(name, delay, ports, configs))
+        for (portIdx in ports) {
+            val port = ports[portIdx]
+            assert(nodeAtEndpoint.get(port) == null) {
+                "Assertion failed: section ${trackSectionPool[port.value].name}.${port.endpoint} is referenced multiple times by a node port"
+            }
+            nodeAtEndpoint.getOrPut(port) { nodeIdx }
```
Comment (eckter) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]:
```suggestion
            assert(nodeAtEndpoint[port] == null) {
                "Assertion failed: section ${trackSectionPool[port.value].name}.${port.endpoint} is referenced multiple times by a node port"
            }
            nodeAtEndpoint[port] = nodeIdx
```

`getOrPut` is a little misleading on the intent here

File: core/kt-osrd-sim-interlocking/src/test/kotlin/fr/sncf/osrd/sim/TestLocation.kt [hunk 2]
```diff
@@ -19,10 +21,23 @@ class TestLocation {
...
             val switchA =
-                movableElement("A", delay = 42L.milliseconds) {
-                    config("a", Pair(TrackNodePortId(0u), TrackNodePortId(1u)))
-                    config("b", Pair(TrackNodePortId(0u), TrackNodePortId(2u)))
-                }
+                node(
+                    "A",
+                    42L.milliseconds,
+                    StaticPool(),
+                    StaticPool(
```
Comment (eckter) [inline core/kt-osrd-sim-interlocking/src/test/kotlin/fr/sncf/osrd/sim/TestLocation.kt]: No port but non-empty config list with node port ids? That doesn't feel right, does it work properly? Or maybe it doesn't matter for the tests?

(the same applies to most test changes)

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 3]
```diff
@@ -535,8 +536,60 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): SimInfraAdapter {
...
     }
 
-    // TODO: remove this stitching between new way of loading track-section info and previous way of
-    // loading network
+    // parse nodes
+    val switchTypeMap =
+        listOf(rjsInfra.switchTypes, RJSSwitchType.BUILTIN_NODE_TYPES_LIST)
+            .flatten()
+            .map { switchType -> Pair(switchType.id, switchType) }
+            .toMap()
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Fancy suggestion from intellij

```suggestion
            .associateBy { switchType -> switchType.id }
```

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 4]
```diff
@@ -535,8 +536,60 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): SimInfraAdapter {
...
+                    builder.getSectionEndpointIdx(
+                        rjsPort.track,
+                        Endpoint.fromEdgeEndpoint(rjsPort.endpoint)
+                    )
+                )
+        }
+        val switchType = switchTypeMap[rjsNode.switchType]
+        assert(switchType != null) {
+            "Assertion failed: Node ${rjsNode.id} references unknown switch-type ${rjsNode.switchType}"
+        }
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: An assertion isn't quite right for this purpose, with default jvm config they're disabled and we would want an actual error here.

We can do something like that instead, it even makes the variable non-null from the start:
```suggestion
        val switchType = switchTypeMap[rjsNode.switchType]
            ?: throw RuntimeException("...")
```

(ideally we'd want a custom error type for *all* error cases, but these errors types are easier to refactor in bulk, later on)

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 5]
```diff
@@ -553,57 +606,16 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): SimInfraAdapter {
...
-                for (edge in edges) {
-                    val endpoint =
-                        if (infra.trackGraph.incidentNodes(edge).nodeU() == node) Endpoint.START
-                        else Endpoint.END
-                    ports.add(port(EndpointTrackSectionId(oldTrackSectionMap[edge]!!, endpoint)))
-                }
-                config("default", Pair(ports[0], ports[1]))
-            }
+        val switchGroups = mutableMapOf<String, TrackNodeConfigId>()
+        val configs = builder.getTrackNodePool().get(nodeNameToIdxMap[oldSwitch.id]!!).configs
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]:
```suggestion
        val configs = builder.getTrackNodePool()[nodeNameToIdxMap[oldSwitch.id]!!].configs
```

File: core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt [hunk 6]
```diff
@@ -134,6 +136,92 @@ open class ComparableChunk(simInfra: RawInfra, chunkId: TrackChunkId) {
     }
 }
 
+open class ComparableSectionEndpoint(val sectionName: String, val endpoint: Endpoint) {
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt]: I don't think we need to make them `open`

But they could be a `data class`, that way we get default `equals` and `hashCode`

**Review by eckter — APPROVED** [review]

File: core/kt-osrd-sim-interlocking/src/test/kotlin/fr/sncf/osrd/sim/TestLocation.kt (hunk 2 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-interlocking/src/test/kotlin/fr/sncf/osrd/sim/TestLocation.kt]: I must admit I didn't dig deeper than "no call to [MovableElementDescriptorBuilder.port](https://github.com/osrd-project/osrd/blob/629612c8fa847ebc2c01226f0fb36830b5dc4766/core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt#L79-L98) was done before" so I kept the same.
It looks like it doesn't matter for the tests, as there are not even track-sections on those tests.
So I went further on removing unused data after your remark in https://github.com/osrd-project/osrd/pull/6799/commits/c4d0a6a9f7a45fa3697f8a392413181cd6cc6e5d

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 5 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Done in https://github.com/osrd-project/osrd/pull/6799/commits/c4d0a6a9f7a45fa3697f8a392413181cd6cc6e5d

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 4 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: I actually chose to throw a (little bit) more specific [InfraHardLoadingError](https://github.com/osrd-project/osrd/pull/6799/commits/c4d0a6a9f7a45fa3697f8a392413181cd6cc6e5d#diff-0335c3a24d409feef2e7c3f5f3173cc4d08928c658e17faa8a0e71384922e6d3R561)

Note that I also moved some error-generating code from [UndirectedInfraBuilder](https://github.com/osrd-project/osrd/pull/6799/commits/c4d0a6a9f7a45fa3697f8a392413181cd6cc6e5d#diff-4042b4f548789226b812e24dd848ee8fcfe851944e921fc46fd4c7627b301426L84-L116) to OSRDError

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 3 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Done in https://github.com/osrd-project/osrd/commit/c4d0a6a9f7a45fa3697f8a392413181cd6cc6e5d

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt (hunk 1 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: Done in https://github.com/osrd-project/osrd/commit/c4d0a6a9f7a45fa3697f8a392413181cd6cc6e5d

File: core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt (hunk 6 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt]: Much less code 👍, done in https://github.com/osrd-project/osrd/pull/6799/commits/e92c7868d8f4b38110d616328fbca5167dd5e72c

**Review by eckter — APPROVED** [review]

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt [hunk 7]
```diff
@@ -84,31 +86,49 @@ class RawInfraFromRjsBuilderImpl : RawInfraBuilder {
...
+        val sectionIdx = getSectionIdx(sectionName)
         return sectionDistanceSortedChunkMap[sectionIdx]
             ?: throw OSRDError.newInfraLoadingError(
                 ErrorType.InfraHardLoadingError,
                 "Accessing sorted chunks from unregistered track-section idx $sectionIdx (name: $sectionName)"
             )
     }
 
-    override fun movableElement(
+    fun getSectionEndpointIdx(sectionName: String, endpoint: Endpoint): EndpointTrackSectionId {
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: I know it's a little bit verbose, but just section doesn't really make it clear what it is. There are other kinds of sections, such as track vacancy detection sections, or neutral sections.

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt [hunk 8]
```diff
@@ -84,31 +86,49 @@ class RawInfraFromRjsBuilderImpl : RawInfraBuilder {
         return sectionDistanceSortedChunkMap
     }
 
+    // TODO remove this accessor once useless in adapter
+    fun getTrackNodePool(): StaticPool<TrackNode, TrackNodeDescriptor> {
+        return trackNodePool
+    }
+
+    private fun getSectionIdx(sectionName: String): TrackSectionId {
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: Same as below, section is an ambiguous shortcut. I'd rather have the full tracksection

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt [hunk 9]
```diff
@@ -252,12 +252,6 @@ class RouteDescriptorImpl(
 ) : RouteDescriptor
 
 interface RestrictedRawInfraBuilder {
-    fun movableElement(
-        name: String,
-        delay: Duration,
-        init: MovableElementDescriptorBuilder.() -> Unit
-    ): TrackNodeId
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt:259]: I'm unsure why this method is gone from the interface. Isn't there a node method now?

File: core/kt-osrd-signaling/src/test/kotlin/fr/sncf/osrd/signaling/BlockBuilderTest.kt [hunk 10]
```diff
@@ -27,13 +28,9 @@ class BlockBuilderTest {
         //  <-- reverse     normal -->
 
         // region build the test infrastructure
-        val builder = RawInfraFromRjsBuilder()
+        val builder = RawInfraFromRjsBuilderImpl()
```
Comment (multun) [inline core/kt-osrd-signaling/src/test/kotlin/fr/sncf/osrd/signaling/BlockBuilderTest.kt:31]: We try not to directly instanciate Impl classes: Impl is a suffix which means this implements some interface. The exact implementation type can be kept private, and a static method can be used to construct it (called something like rawInfraFromRJSBuilder)

**Review by Erashin — APPROVED** [review]

File: core/kt-osrd-signaling/src/test/kotlin/fr/sncf/osrd/signaling/BlockBuilderTest.kt (hunk 10 above)
Comment (bougue-pe) [reply core/kt-osrd-signaling/src/test/kotlin/fr/sncf/osrd/signaling/BlockBuilderTest.kt:31]: As discussed, the plan is to drop completely RawInfraBuilder interface (at least RestrictedRawInfraBuilder).
Respecting the previous names for now was a way to follow work (but not explicit on the ambition to drop interface).
This is tracked in parent issue, and will be done in a separate PR (maybe after ongoing work on detection zones).

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt (hunk 9 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraBuilder.kt:259]: As discussed, the goal is to drop all the RawInfraBuilder interface.
Dropping the unused methods from interface iteratively (when unused in the new Builder).
Naming `node()` the new method seemed better to me.

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt (hunk 7 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: Verbose is fine 👍 
Done in a larger dedicated commit: https://github.com/osrd-project/osrd/pull/6799/commits/16c474bcec07c1972c811b5e5ba40eba26c47666 (and a little change amended for this line in https://github.com/osrd-project/osrd/pull/6799/commits/deed93c0c8481d17cbdf197dfb15cf5b8f690419#diff-0d226bd7ab1d615284f3c45cc939f9e5db6d8a3cf4f86837ada3b24e85369367R125)

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt (hunk 8 above)
Comment (bougue-pe) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/RawInfraFromRjsBuilder.kt]: Done in https://github.com/osrd-project/osrd/pull/6799/commits/16c474bcec07c1972c811b5e5ba40eba26c47666

**Review by multun — APPROVED** [review]

**Pull request merged**